keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+* Translate pod-level resource specifications to container-level specifications.
+
+* Delete existing workloads and recreate them with container-level specifications.
+
+* Drain the nodes one by one to prevent the new pods from being scheduled on them.
+
+* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
+fields.
+
+* Downgrade the control plane components.
+
+* Uncordon the nodes, making it available for scheduling again.
+
+
+### Version Skew Strategy
+
+Pod-level resource specification is an opt-in feature. For this feature to work correctly, it must be enabled
+in all parts of the cluster (scheduler, API server, kubelet). However, if a node is running an older version
+of kubelet that doesn't support this feature, it will ignore the pod-level resouce specifications and continue
+to use the aggregated container-level values (backward compatible). Applications must be prepared for this version
+skew implementation.
+
+For users wanting to use this feature, it is user's responsibility to use the correct version of kube-scheduler,
+kube-apiserver and kubelet otherwise the feature will not work as expected.
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLevelResources
+  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
+    The workloads need to restart to disable the feature.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? No. Only the workloads need to restart.
+
+###### Does enabling the feature change any default behavior?
+
+No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Once the feature gates are disabled, pods created with pod level resource spec need to be deleted and recreated with only container level resources.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Rollbacks will require disabling the feature, and recreating the pods that were using this new feature.
```

Comment:
it is rare that we require pods recreation on feature disablement. Also the question is asking what will happen with the running pods when feature was flipped back and forth

Reply:
> Also the question is asking what will happen with the running pods when feature was flipped back and forth

The existing pods with the feature enabled should continue run without issues assuming they are already running. If they haven't been created on the node, kubelet would ignore the pod-level resource spec, and go ahead and set up the pods/containers with the container spec. In a way, this would be working as intended.

Reply:
Note that we don't support rollback of alpha features.  So the question here is really:  "Once this feature goes to beta, and the previous version has the implementation and feature gate, what happens during rollback/reenablement?"   Ideally, the feature continues to work as expected when rolled back, but if the previous release did not have the feature gate enabled-by-default, no additional pods may be created with the feature, but once we reenable the feature both existing pods continue to work and new pods can be created that use the feature.